File: csharp/feature-complete/Controllers/AuthController.cs
using Microsoft.AspNetCore.Mvc;
using FluentValidation;
using feature_complete.Models;
using feature_complete.Services;
using feature_complete.Validators;

namespace feature_complete.Controllers
{
    [ApiController]
    [Route("api/auth")]
    public class AuthController : ControllerBase
    {
        private readonly AuthService _authService;
        private readonly RegisterRequestValidator _registerValidator;
        private readonly LoginRequestValidator _loginValidator;

        public AuthController(
            AuthService authService,
            RegisterRequestValidator registerValidator,
            LoginRequestValidator loginValidator)
        {
            _authService = authService;
            _registerValidator = registerValidator;
            _loginValidator = loginValidator;
        }

        [HttpPost("register")]
        public IActionResult Register([FromBody] RegisterRequest request)
        {
            try
            {
                var validationResult = _registerValidator.Validate(request);
                if (!validationResult.IsValid)
                {
                    return FormatValidationErrors(validationResult);
                }

                var result = _authService.Register(request.Username, request.Password);
                
                if (result == null)
                {
                    return Conflict(new { error = "Username already exists" });
                }

                return StatusCode(201, result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = "Registration failed", message = ex.Message });
            }
        }

        [HttpPost("login")]
        public IActionResult Login([FromBody] LoginRequest request)
        {
            try
            {
                var validationResult = _loginValidator.Validate(request);
                if (!validationResult.IsValid)
                {
                    return FormatValidationErrors(validationResult);
                }

                var result = _authService.Authenticate(request.Username, request.Password);
                
                if (result == null)
                {
                    return Unauthorized(new { error = "Invalid credentials" });
                }

                return Ok(result);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = "Login failed", message = ex.Message });
            }
        }

        [HttpGet("profile")]
        public IActionResult GetProfile()
        {
            try
            {
                var userId = GetCurrentUserId();
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { error = "Authentication required" });
                }

                var profile = _authService.GetProfile(userId);
                
                if (profile == null)
                {
                    return NotFound(new { error = "User not found" });
                }

                return Ok(profile);
            }
            catch (Exception ex)
            {
                return StatusCode(500, new { error = "Failed to get profile", message = ex.Message });
            }
        }

        private string? GetCurrentUserId()
        {
            return HttpContext.Items["UserId"] as string;
        }

        private IActionResult FormatValidationErrors(FluentValidation.Results.ValidationResult validationResult)
        {
            var errors = validationResult.Errors.Select(e => new
            {
                field = e.PropertyName.ToLower(),
                message = e.ErrorMessage
            }).ToList();

            return BadRequest(new { errors });
        }
    }
}

File: typescript/feature-complete/src/controllers/authController.ts
import { Response } from 'express';
import { db } from '../models/database';
import { AuthRequest } from '../middleware/auth';
import { CreateUserDTO, LoginDTO, AuthResponse } from '../models/user.model';
import { v4 as uuidv4 } from 'uuid';
import bcrypt from 'bcryptjs';
import { generateToken } from '../middleware/auth';

export class AuthController {
  static register(req: AuthRequest, res: Response): void {
    const { username, email, password } = req.body as CreateUserDTO;
    const id = uuidv4();
    const hashedPassword = bcrypt.hashSync(password, 10);

    db.run(
      `INSERT INTO users (id, username, email, password) VALUES (?, ?, ?, ?)`,
      [id, username, email, hashedPassword],
      function(err) {
        if (err) {
          if (err.message.includes('UNIQUE constraint')) {
            res.status(400).json({ error: 'Username or email already exists' });
            return;
          }
          res.status(500).json({ error: 'Failed to register user' });
          return;
        }

        const user = { id, username, email, createdAt: new Date().toISOString() };
        const token = generateToken({ id, email, username });

        res.status(201).json({ user, token } as AuthResponse);
      }
    );
  }

  static login(req: AuthRequest, res: Response): void {
    const { email, password } = req.body as LoginDTO;

    db.get('SELECT * FROM users WHERE email = ?', [email], (err, user: any) => {
      if (err) {
        res.status(500).json({ error: 'Failed to login' });
        return;
      }

      if (!user) {
        res.status(401).json({ error: 'Invalid credentials' });
        return;
      }

      const validPassword = bcrypt.compareSync(password, user.password);
      if (!validPassword) {
        res.status(401).json({ error: 'Invalid credentials' });
        return;
      }

      const safeUser = { id: user.id, username: user.username, email: user.email };
      const token = generateToken(safeUser);

      res.json({ user: safeUser, token } as AuthResponse);
    });
  }

  static getProfile(req: AuthRequest, res: Response): void {
    const user = req.user;
    if (!user) {
      res.status(404).json({ error: 'User not found' });
      return;
    }

    db.get('SELECT id, username, email, createdAt FROM users WHERE id = ?', [user.id], (err, userData: any) => {
      res.json(userData);
    });
  }
}
